ERCS/erc-7677.md

```diff
@@ -0,0 +1,350 @@
+author: Lukas Rosario (@lukasrosario), Dror Tirosh (@drortirosh), Wilson Cusack (@wilsoncusack), Kristof Gazso (@kristofgazso), Hazim Jumali (@hazim-j)
+discussions-to: https://ethereum-magicians.org/t/erc-7677-paymaster-web-service-capability/19530
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-04-03
+requires: 4337, 5792
+---
+
+## Abstract
+
+With [EIP-5792](./eip-5792.md), apps can communicate with wallets about advanced features via capabilities. This proposal defines a capability that allows apps to request that [ERC-4337](./eip-4337.md) wallets communicate with a specified paymaster web service. To support this, we also define a standardized API for paymaster web services.
+
+## Motivation
+
+App developers want to start sponsoring their users' transactions using paymasters. Paymasters are commonly used via web services. However, there is currently no way for apps to tell wallets to communicate with a specific paymaster web service. Similarly, there is no standard for how wallets should communicate with these services. We need both a way for apps to tell wallets to communicate with a specific paymaster web service and a communication standard for wallets to do so.
+
+## Specification
+
+One new [EIP-5792](./eip-5792.md) wallet capability is defined. We also define a standard interface for paymaster web services as a prerequisite.
+
+### Paymaster Web Service Interface
+
+We define two JSON-RPC methods to be implemented by paymaster web services.
+
+#### `pm_getPaymasterStubData`
+
+Returns stub values to be used in paymaster-related fields of an unsigned user operation for gas estimation. Accepts an unsigned user operation, entrypoint address, chain id, and a context object. Paymaster service providers can define fields that app developers should use in the context object. 
+
+This method MAY return paymaster-specific gas values if applicable to the provided EntryPoint version. For example, if provided with EntryPoint v0.7, this method MAY return `paymasterVerificationGasLimit` and `paymasterPostOpGasLimit` values. The wallet SHOULD use these provided gas values when submitting the `UserOperation` to a bundler for gas estimation.
+
+##### `pm_getPaymasterStubData` RPC Specification
+
+Note that the user operation parameter does not include a signature, as the user signs after all other fields are populated.
+
+```typescript
+// [userOp, entryPoint, chainId, context]
+type GetPaymasterStubDataParams = [
+  // Below is specific to Entrypoint v0.6 but this API can be used with other entrypoint versions too
+  {
+    sender: `0x${string}`;
+    nonce: `0x${string}`;
+    initCode: `0x${string}`;
+    callData: `0x${string}`;
+    callGasLimit: `0x${string}`;
+    verificationGasLimit: `0x${string}`;
+    preVerificationGas: `0x${string}`;
+    maxFeePerGas: `0x${string}`;
+    maxPriorityFeePerGas: `0x${string}`;
+  }, // userOp
+  `0x${string}`, // EntryPoint
```

Comment:
it is better to have named parameters instead of indexed.
e.g. 
```js
{ userOp: {}, chainId: "0x..", entryPoint: "0x...", version: "0.7", context: {} }
```
And also be better to add "version", instead of rely on entrypoint address

Reply:
curious about your reasoning here. i like the clarity of named arguments but most ethereum-related JSON-RPC methods i can think of use positional arguments. is there a reason you think we should stray from that loose convention other than clarity?

Reply:
same question for the `version` arg, is that just for clarity?